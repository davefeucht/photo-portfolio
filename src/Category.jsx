/*****************
* Category component displays the posts for a particular category.
* Props:
*  id: PropTypes.number,
*  name: PropTypes.string,
*  site: PropTypes.string,
*  showAllPosts: PropTypes.bool,
*  clickCategory: PropTypes.func
*****************/

import React from "react";
import axios from "axios";
import Posts from "./Posts.jsx";
import Post from "./Post.jsx";
import PropTypes from "prop-types";

class Category extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      categoryPost: {},
      errorMsg: ""
    };
  }

  _getCategoryPost(categoryId) {
    let getCategoryPostURI = "http://" + this.props.site + "/wp-json/wp/v2/posts?categories=" + categoryId;
    axios.get(getCategoryPostURI)
      .then(res => {
        if(res.data[0] !== undefined) {
          const categoryPost = res.data[0];
          this.setState({categoryPost});
        }
      }, error => {
        const errorMsg = "Did not work: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  }

  _openCategory() {
    this.setState({showAllPosts: !this.state.showAllPosts}); 
    this.props.clickCategory(this.props.id, this.props.name);
  }

  componentWillMount() {
    this._getCategoryPost(this.props.id);
  }

  render() {
    let postBody = "";
    let categoryTitle = <div className="category-title-clickable">{this.props.name}</div>;
    let categorySubtitle = "";
    if(!this.props.showAllPosts) {
      if(!(Object.keys(this.state.categoryPost).length === 0 && this.state.categoryPost.constructor === Object)) {
        postBody = <Post key={this.state.categoryPost.id.toString()} category={this.props.id} categoryName={this.props.name} title={this.state.categoryPost.title.rendered} id={this.state.categoryPost.id} image={this.state.categoryPost.featured_media} context="category-image" site={this.props.site} />;
      }
    }
    else {
      categoryTitle = <div className="category-title">{this.props.name}</div>;
      categorySubtitle = <div className="category-subtitle">Back to Categories</div>; 
      postBody = <Posts site={this.props.site} category={this.props.id} />;
    }
    return ( 
      <div className="category" onClick={this._openCategory.bind(this)}>
        <div>
          {categoryTitle}
          {categorySubtitle}
        </div>
        {postBody}
      </div>
    );
  }

}

Category.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  site: PropTypes.string,
  showAllPosts: PropTypes.bool,
  clickCategory: PropTypes.func
};

export default Category;
