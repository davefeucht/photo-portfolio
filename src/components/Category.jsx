/*****************
* Category component displays the posts for a particular category.
* Props:
*  id: PropTypes.number,
*  name: PropTypes.string,
*  site: PropTypes.string,
*  clickCategory: PropTypes.func
*****************/

import React from "react";
import axios from "axios";
import Posts from "./Posts.jsx";
import CategoryHeader from "./CategoryHeader.jsx";
import StyledCategory from "./styledComponents/StyledCategory.jsx";
import PropTypes from "prop-types";

export default class Category extends React.Component {

  static propTypes = {
    id: PropTypes.number,
    name: PropTypes.string,
    site: PropTypes.string,
    clickCategory: PropTypes.func
  };

  state = {
    categoryPost: {},
    errorMsg: ""
  };

  //Function to get the 'main' post for the category
  _getCategoryPost(categoryId) {
    let getCategoryPostURI = "http://" + this.props.site + "/wp-json/wp/v2/posts?categories=" + categoryId;
    axios.get(getCategoryPostURI)
      .then(res => {
        if(res.data[0] !== undefined) {
          const categoryPost = res.data[0];
          this.setState({categoryPost});
        }
      })
      .catch(error => {
        const errorMsg = "Could not get posts for this category: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  }

  //When the component is about to mount, get the main post for the category
  componentWillMount() {
    this._getCategoryPost(this.props.id);
  }

  render() {
    return ( 
      <StyledCategory>
        <CategoryHeader id={this.props.id} name={this.props.name} clickCategory={this.props.clickCategory.bind(this)} />
        <Posts site={this.props.site} category={this.props.id} />
      </StyledCategory>
    );
  }
}