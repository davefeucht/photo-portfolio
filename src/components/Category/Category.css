.category {
  min-height: 100%;
  display: grid;
  grid-template-columns: 100%;
  grid-template-areas: "category-header" "category-content";
  grid-template-rows: .25fr 4fr;
}