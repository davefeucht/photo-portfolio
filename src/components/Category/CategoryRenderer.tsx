import Posts from 'components/Posts/Posts';
import SectionHeader from 'components/SectionHeader/SectionHeader';
import { observer } from 'mobx-react';
import * as React from 'react';
import { Outlet } from 'react-router-dom';
import { Post, ScreenInfo } from 'utils/types';

interface Props {
    categoryId: number;
    maxItemsPerPage: number;
    categoryName?: string;
    currentCategoryPosts: Post[];
    screenInfo: ScreenInfo;
}

const CategoryRenderer: React.FC<Props> = ({
    categoryId,
    maxItemsPerPage,
    categoryName,
    currentCategoryPosts,
    screenInfo
}) => {
    return (
        <div className="category">
            <SectionHeader title={categoryName ?? ""} />
            <Posts
                maxItemsPerPage={maxItemsPerPage}
                screenInfo={screenInfo}
                categoryId={categoryId}
                currentCategoryPosts={currentCategoryPosts}
            />
            <Outlet />
        </div>
    );
};

export default observer(CategoryRenderer);
