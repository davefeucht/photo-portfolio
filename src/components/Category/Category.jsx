/*****************
* Category component displays the posts for a particular category.
*****************/

import React from 'react';
import { observer } from 'mobx-react';
import { 
  useParams,
} from 'react-router-dom';
import Posts from '../Posts/Posts.jsx';
import CategoryHeader from '../CategoryHeader/CategoryHeader.jsx';
import './Category.css';

const Category = observer(({ stateStore, api }) => {
  let { categoryId } = useParams();

  api.getPosts(categoryId);
  api.getCategoryInfo(categoryId);

  return ( 
    <div className="category">
      <CategoryHeader stateStore={stateStore} categoryData={stateStore.currentCategoryData}/>
      <Posts stateStore={stateStore} categoryId={categoryId} api={api} />
    </div>
  );
});

Category.displayName = 'Category';

export default Category;