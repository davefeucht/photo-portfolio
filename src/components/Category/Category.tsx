/** ***************
* Category component displays the posts for a particular category.
**************** */

import './Category.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect } from 'react';
import {
    Outlet,
    useParams
} from 'react-router-dom';

import { getCategoryInfo, getPosts, getPostThumbnail } from '../../utils/Api';
import {
    Category as CategoryState,
    ImageData,
    Post as PostState,
    ScreenInfo,
    SiteInfo
} from '../../utils/types';
import Posts from '../Posts/Posts';
import SectionHeader from '../SectionHeader/SectionHeader';

interface CategoryProps {
    maxItemsPerPage: number,
    siteInfo: SiteInfo,
    screenInfo: ScreenInfo,
    currentCategoryPosts: PostState[],
    currentCategoryData: CategoryState,
    setCategoryPosts: (a: PostState[]) => void,
    setCategoryData: (a: CategoryState) => void,
    setThumbnailImageUrl: (a: ImageData) => void
}

const Category: React.FC<CategoryProps> = ({
    maxItemsPerPage,
    siteInfo,
    screenInfo,
    currentCategoryPosts,
    currentCategoryData,
    setCategoryPosts,
    setCategoryData,
    setThumbnailImageUrl
}) => {
    const { categoryId = '' } = useParams();

    useEffect(() => {
        if (categoryId) {
            getPosts(parseInt(categoryId), siteInfo.siteUrl)
                .then((posts: PostState[]) => {
                    setCategoryPosts(posts);
                    currentCategoryPosts.forEach((post, index) => {
                        getPostThumbnail(post.featured_media, siteInfo.siteUrl)
                            .then((thumbUrl: string) => {
                                setThumbnailImageUrl({ post_index: index, image_url: thumbUrl });
                            });
                    });
                });
            getCategoryInfo(parseInt(categoryId), siteInfo.siteUrl)
                .then((categoryInfo: CategoryState) => {
                    setCategoryData(categoryInfo);
                });
        }
    }, [categoryId]);

    return (
        <div className="category">
            <SectionHeader title={currentCategoryData.name} />
            <Posts
                maxItemsPerPage={maxItemsPerPage}
                screenInfo={screenInfo}
                categoryId={parseInt(categoryId)}
                currentCategoryPosts={currentCategoryPosts}
            />
            <Outlet />
        </div>
    );
};

Category.displayName = 'Category';

export default observer(Category);
