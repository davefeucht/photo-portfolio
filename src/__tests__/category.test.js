import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Category from '../components/Category/Category';
import stateStore from '../StateStore/store';

const posts = [
    {
        id: 887,
        featured_media: 142,
        categories: [220]
    },
    {
        id: 824,
        featured_media: 820,
        categories: [220]
    }
];

const category = {
    id: 35,
    count: 3,
    description: 'Category of things',
    link: '',
    name: 'Some category',
    slug: 'smcg',
    taxonomy: '',
    parent: 23,
    meta: {}
};

jest.mock('../utils/Api', () => ({
    getPosts: () => Promise.resolve(posts),
    getCategoryInfo: () => Promise.resolve(category)
}));

test('Category displays', () => {
  const store = new stateStore();
  let component;
  let tree;
  component = renderer.create(
      <MemoryRouter initialEntries={['/category/35']} >
          <Category stateStore={store}></Category>
      </MemoryRouter>
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
