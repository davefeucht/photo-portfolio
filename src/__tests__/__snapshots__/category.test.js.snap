// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category displays 1`] = `
<div
  className="category"
>
  <div
    className="section-header"
  >
    <div
      className="title"
    />
  </div>
  <div
    className="posts"
  />
</div>
`;
